cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillController.java

```diff
@@ -449,37 +441,160 @@ public void onSuccess(V1Status v1Status, int statusCode, Map<String, List<String
+                                                   try {
+                                                     coreApi.deleteCollectionNamespacedPersistentVolumeClaimAsync(
+                                                       kubeNamespace, null, null, null, null, null, getLabelSelector(),
+                                                       null, null, null, null, null, null, null,
+                                                       createCallbackFutureAdapter(resultFuture, r -> name));
+                                                   } catch (ApiException e) {
+                                                     completeExceptionally(resultFuture, e);
+                                                   }
+                                                 }
+                                               });
+
+    } catch (ApiException e) {
+      completeExceptionally(resultFuture, e);
+    }
+    return resultFuture;
+  }
+
+  /**
+   * Cleans up the resources for deployment and stateful set and returns the future.
+   */
+  private CompletableFuture<KubeTwillController> cleanupResources() {
+    CompletableFuture<KubeTwillController> resultFuture;
+    CompletionStage<String> completionStage;
+    resultFuture = new CompletableFuture<>();
+    // delete the resource
+    completionStage = deleteResource();
+    completionStage.whenComplete((name, t) -> {
+      if (t != null) {
+        resultFuture.completeExceptionally(t);
+      } else {
+        resultFuture.complete(KubeTwillController.this);
+      }
+    });
+    // If the deletion was successful, reflect the status into the completion future
+    // If the deletion was failed, don't change the completion state so that caller can retry
+    resultFuture.thenAccept(completion::complete);
+    return resultFuture;
+  }
+
+  /**
+   * Cleans up the job and returns future with correct job status.
+   */
+  private CompletableFuture<KubeTwillController> cleanupJob() {
+    // get job status
+    CompletionStage<String> completionStage = getJobCompletionStatus();
+    completionStage.whenComplete((name, t) -> {
+      if (t != null) {
+        completion.completeExceptionally(t);
+      } else {
+        completion.complete(KubeTwillController.this);
+      }
+    });
+    return completion;
+  }
+
+  /**
+   * Returns job completion status.
+   */
+  private CompletionStage<String> getJobCompletionStatus() {
+    CompletableFuture<String> resultFuture = new CompletableFuture<>();
+    String name = meta.getName();
+    // In case we fail to get job status due to some transient issue, retry 5 times. There is no need to retry for too
+    // long otherwise corresponding program may appear to be stuck while retrying to get job status.
+    // If we exhaust retries, complete future exceptionally. Though in normal scenario, retries is not expected
+    // to be performed.
+    int retryCount = 5;
```

Comment:
Why need retry at all since we have the job cleaner?

Reply:
In case its just transient error. But we can remove it.

Reply:
Removed the retries